### PR TITLE
Change Hardware return type to enum class

### PR DESCRIPTION
**Pull request by destogl**
As [commented](https://github.com/ros-controls/ros2_control/pull/101#discussion_r445662475) on #101 by @bmagyar

### CODE REVIEW

File: hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp [hunk 1]
```diff
@@ -17,9 +17,21 @@
 
 namespace hardware_interface
 {
-using hardware_interface_ret_t = int;
-static constexpr hardware_interface_ret_t HW_RET_OK = 0;
-static constexpr hardware_interface_ret_t HW_RET_ERROR = 1;
+enum class HardwareReturnType
+{
+  HW_RET_OK = 0,
```
Comment (bmagyar) [inline hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp:22]: I'd go as far as streamlining these value identifiers too as the type is now mandatory to be written out in full so the names don't need to encode it
```suggestion
  OK = 0,
```

File: hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp [hunk 2]
```diff
@@ -17,9 +17,21 @@
 
 namespace hardware_interface
 {
-using hardware_interface_ret_t = int;
-static constexpr hardware_interface_ret_t HW_RET_OK = 0;
-static constexpr hardware_interface_ret_t HW_RET_ERROR = 1;
+enum class HardwareReturnType
```
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp:20]: this should inherit from `std::uint8_t` or something

**Comment by bmagyar** [comment]
Both `ros2_controllers` and `gazebo_ros2_control` will be affected. Once this PR is final and merged I'm happy to go in and fix up the rest

**Comment by Karsten1987** [comment]
I would propose to open a connected PR for `ros2_controllers` and run CI for all PRs.

**Comment by bmagyar** [comment]
@destogl @Karsten1987 I'll go ahead and address the comments on this PR and open the ros2_controllers one

**Comment by bmagyar** [comment]
So testing it against controller revealed a couple of cases where hardware interface return types were used instead of controller interface return types. the funny thing is that value-wise they weren't compatible (1 for success, 0 for error and vice-versa).

I took the liberty of applying the same changes to controller_interface_ret_t

**Comment by bmagyar** [comment]
I kept `hardware_interface_ret_t` for compliance with pending PR-s but removed it's uses in the framework

File: hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp (hunk 2 above)
Comment (bmagyar) [reply hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp]: roger roger

File: hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp [hunk 3]
```diff
@@ -15,11 +15,25 @@
...
+  OK = 0,
+  ERROR = 1,
+
+  CLAIMED_ERROR = 10,
+  ALREADY_CLAIMED = 11,
+  NOT_CLAIMED = 11,
+  UNATHORIZED_UNCLAIM = 13,
+  NON_CLAIMED_WRITE = 15,
+
+  CAN_NOT_READ = 20,
```
Comment (Karsten1987) [inline hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp:33]: Where are these values being used?

Comment (destogl) [reply hardware_interface/include/hardware_interface/types/hardware_interface_return_values.hpp:33]: Currently nowhere... but I am planing to use them in actuators. Should I remove this Andrade when it is used?

**Pull request merged**